### DISCUSSION TITLE
VRPTW problem with nodes are given as time lapse.

### DISCUSSION THREAD

**Discussion by gyang274** [google/or-tools#3349] [Routing (and legacy CP) questions]
Hi, I am working on a VRPTW problem with an additional constraints on "knowing the next node after a specific time". The idea is to simulate a delivery service where the next locations are given dynamically.

---- EX1
For example, a vehicle is available at 4 pm at the starting location X, 1st order is placed at 4:30 pm at location A, 2nd order placed at 6:00 pm at location B. The travel time from X to A, X to B and A to B are all 10 min, the service time at A and B are 10 min.

In VRPTW, vehicle will go X (4:00 pm) -> A (4:30 pm) -> B (6:00 pm). We want to add the constraints so that vehicle won't leave X until knowing A is placed, so it should start driving to A at 4:30 pm, and arrive A at 4:40 pm. Similarly, vehicle should start heading to B at 6:00 pm, and arrive at 6:10 pm.

So, we aim to get X (4:00 pm) -> A (4:40 pm) -> B (6:10 pm).

---- EX2
Another example to illustrate the difference, vehicle is available at 4 pm at the starting location X, 1st order is placed at 4:30 pm at location A, 2nd order placed at 4:45 pm at location C. The travel time from X to A, X to C and A to C are all 10 min, the service time at A and C are 10 min.

We aim to get X (4:00 pm) -> A (4:40 pm) -> C (5:00 pm), start driving from X to A at 4:30 pm, once A is known (order A being placed at) 4:30 pm, arrive at A at 4:40, do the service for 10 min, being available at 4:50 pm. At 4:50 pm, B is being placed, so drive to B immediately and arrive at 5:00 pm.

Looking forward to hearing your advice.

**Comment by Mizux** [comment]
Add constraint on span of the route ?
https://github.com/google/or-tools/blob/2115d65c35181366dd80733eac979d3bcfa53146/ortools/constraint_solver/routing.h#L2712-L2719

Also you can use:
https://github.com/google/or-tools/blob/2115d65c35181366dd80733eac979d3bcfa53146/ortools/constraint_solver/samples/vrp_time_windows.py#L159-L166
to shrink the vehicle span

**Comment by gyang274** [comment]
HI Mizux, thank you for your prompt reply. And I learned a lot from your posts on github and stack-overflows, thanks.

I am not sure how route span cost can help to specify "staying until node become known".

In my mind, I need to constraint like:

routing.solver().Add(
  time_dimension.CumulVar(i) + time_dimension.SlackVar(i) >= data['order_placed'][j] if j == next(i)
)

The question is I can only get the next(i) after finding the solution, e.g., j == solution.Value(routing.NextVar(i)).

How to specify it in a constraints?

Btw, I edited the example for make it more clear.

**Comment by Mizux** [comment]
got it, need to think of it, seems you want to simulate order coming little by little than planning everything ahead of time.

ps: near off topic but just  in case https://gist.github.com/Mizux/0bd1daadc4e64130b62492c942f163d5

**Reply by gyang274** [reply]
Yes, exactly. The orders are being placed one by one over the night by customers, we want to see what is the best routing to minimize the drive time or order-to-eat time. At this moment, the excise is retrospective, so we know when/where the orders are being place at the beginning, but we want to keep the information from the algorithm until it is revealed.

**Comment by gyang274** [comment]
following:  https://gist.github.com/Mizux/0bd1daadc4e64130b62492c942f163d5, and the example above how about:

X (4:00 pm) -> A (4:30 pm) -> B (6:00 pm), time in parenthesis are order being place, known to the algorithm.

For node A, augment node with demand quantity -1. And, create proxies node XA, BA, pickup quantity 1, time window XA 4:40 pm, BA 4:40 pm. Cost of XA-A and BA-A are 0, and infinity from XA or BA to other nodes.

This enforce to go XA or BA before going A, distance is 0, the time on XA or BA ensures it arrived A after it being revealed.

The problem is that this will make the problem from N nodes to about N^2 / 2 nodes, where N is the number of orders.
